### ISSUE TITLE
High Disk Usage / Deleting Resources / Garbage Collection

### ISSUE THREAD

**Issue by dottodot** [flynn/flynn#2658]
Not sure if this is normal but when I run `df -h` the disk usage seems pretty high considering I only have 2 apps and the output is

```
Filesystem                                          Size  Used Avail Use% Mounted on
udev                                                483M  4.0K  483M   1% /dev
tmpfs                                                99M  768K   98M   1% /run
/dev/dm-1                                            37G   29G  7.1G  80% /
none                                                4.0K     0  4.0K   0% /sys/fs/cgroup
none                                                5.0M     0  5.0M   0% /run/lock
none                                                494M  1.8M  492M   1% /run/shm
none                                                100M     0  100M   0% /run/user
/dev/sda1                                           464M   69M  367M  16% /boot
flynn-default/91db808a-7cdf-4c5c-9c5f-9e4cbdd81311   17G  128K   17G   1% /var/lib/flynn/volumes/zfs/mnt/91db808a-7cdf-4c5c-9c5f-9e4cbdd81311
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-f9954677-18c7-4a79-b744-83edeb389579
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-ac8ca909-7cf8-4192-b8d4-db709ec8d7e7
flynn-default/a1673446-dcd2-440e-b99c-ad1cf51c083d   20G  3.0G   17G  16% /var/lib/flynn/volumes/zfs/mnt/a1673446-dcd2-440e-b99c-ad1cf51c083d
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-56454aeb-011f-4987-8cec-5c0ebd2e68f9
flynn-default/3557e8d9-1ad5-4f10-93fa-6831efb509b2   20G  2.9G   17G  16% /var/lib/flynn/volumes/zfs/mnt/3557e8d9-1ad5-4f10-93fa-6831efb509b2
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-7c8b6e28-1d6b-4b71-aeac-b974aa1985db
flynn-default/328338c5-efd4-484d-8559-ec7d82e6fb8c   20G  2.9G   17G  16% /var/lib/flynn/volumes/zfs/mnt/328338c5-efd4-484d-8559-ec7d82e6fb8c
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-ad8a7eb9-f3b4-4b16-a97d-6f7187b85882
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-e3bd6864-e5a8-4f00-baf8-e8393da2bd66
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-b9041883-a73d-4165-a12d-aecc6341beaf
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-54893196-e8e3-4481-88aa-f6440ac2c161
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-1691f884-1f21-41bf-ac93-9881a1c283ff
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-1ec59911-bc3c-4828-8e44-b1b22e0f57db
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-079966ce-02a2-4474-b3e6-d933f8ab4bef
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-63c8eb46-f991-43e0-b66d-465c164c21f7
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-6d8889db-0507-4e60-96c5-630c061472c5
flynn-default/47927988-6177-49e7-8022-a50089c53730   17G     0   17G   0% /var/lib/flynn/volumes/zfs/mnt/47927988-6177-49e7-8022-a50089c53730
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-ae40f8fc-3cda-4554-ba3e-d8b26aa8d6b8
flynn-default/befd25e7-7688-4423-b500-6923e885a70f   17G     0   17G   0% /var/lib/flynn/volumes/zfs/mnt/befd25e7-7688-4423-b500-6923e885a70f
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-3f83a628-d252-471e-ae85-0a65598b0b8a
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-a419a7c7-9fb1-488a-a4cb-2920971e2a67
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-5f7f0aba-2501-4577-b485-d2e38bcd100f
flynn-default/3e3dd044-2d01-4ddc-b743-448bc0b10cc3   17G     0   17G   0% /var/lib/flynn/volumes/zfs/mnt/3e3dd044-2d01-4ddc-b743-448bc0b10cc3
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-cf7519f5-1fdb-4418-ac92-25d2646a1a21
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-1bca5559-4c30-43c3-be8e-244f744bcd2a
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-bf65b5eb-86eb-4100-8bbe-cdb83e75151c
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-296e663b-c04b-4780-a191-23467f1eb9df
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-a5b9ebeb-a10a-4905-9151-494dda987dfe
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-6279c3d5-dd84-4fcc-a3e4-daede3f786e1
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-5f83c81e-3dff-4aa9-9ac4-1a45bf465f31
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-db5669cc-2a95-42ea-85af-85b12dbfb765
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-05596082-8838-47c9-b262-eb3978569d35
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-a4968693-1448-43dc-9c5d-7fd659e4ed3d
none                                                 37G   29G  7.1G  80% /var/lib/docker/aufs/mnt/tmp-cluster1-fd2babf0-0e3f-4bdb-a424-ecc3dcce0c55
```

**Comment by titanous** [comment]
This is normal, we are in the process of adding garbage collection for app release images.

**Comment by dottodot** [comment]
OK. Does this mean I likely to run out of disk space and is there anything I can do to clean up until this is added?

**Comment by titanous** [comment]
Post the out put of `sudo zfs list`, it will have more accurate usage numbers.

**Comment by dottodot** [comment]
OK so I now have a new cluster with 20GB of separate storage setup as per the Storage documentation and as a bit of a test I pushed several updates (about 13) and sure enough the disk filled up at which point I can no longer push any more updates.

The output from `sudo zfs list` is 

```
NAME                                                 USED  AVAIL  REFER  MOUNTPOINT
flynn-default                                       17.9G  15.2M    19K  /flynn-default
flynn-default/064b812c-faa3-4e48-857b-e222812dabb3    19K  15.2M    19K  /var/lib/flynn/volumes/zfs/mnt/064b812c-faa3-4e48-857b-e222812dabb3
flynn-default/0ffd1db6-3721-45d7-a1f7-cc697a68ffc1    19K  15.2M    19K  /var/lib/flynn/volumes/zfs/mnt/0ffd1db6-3721-45d7-a1f7-cc697a68ffc1
flynn-default/1a297a2c-bb37-4f36-bf85-cfbc43f0ea9a    19K  15.2M    19K  /var/lib/flynn/volumes/zfs/mnt/1a297a2c-bb37-4f36-bf85-cfbc43f0ea9a
flynn-default/588f0011-8c9d-415a-8e14-5e73702c7f62    19K  15.2M    19K  /var/lib/flynn/volumes/zfs/mnt/588f0011-8c9d-415a-8e14-5e73702c7f62
flynn-default/65b08c5c-bd4f-41fa-9a2f-ecf267adb7a3    20K  15.2M    20K  /var/lib/flynn/volumes/zfs/mnt/65b08c5c-bd4f-41fa-9a2f-ecf267adb7a3
flynn-default/983169a9-2a1c-4696-a4e1-d98163cf6582   148K  15.2M   148K  /var/lib/flynn/volumes/zfs/mnt/983169a9-2a1c-4696-a4e1-d98163cf6582
flynn-default/c04141c2-4c22-49d1-8453-f2746f22fed0    19K  15.2M    19K  /var/lib/flynn/volumes/zfs/mnt/c04141c2-4c22-49d1-8453-f2746f22fed0
flynn-default/cd252285-a7ed-4f2e-af98-78663ca85ca5    19K  15.2M    19K  /var/lib/flynn/volumes/zfs/mnt/cd252285-a7ed-4f2e-af98-78663ca85ca5
flynn-default/dccddfd5-fe35-4ee6-bb17-c260439794de  17.9G  15.2M  17.9G  /var/lib/flynn/volumes/zfs/mnt/dccddfd5-fe35-4ee6-bb17-c260439794de
```
This looks to be a similar issue https://github.com/flynn/flynn/issues/38

So how can I free disk space as I can't just keep creating new clusters every time happen or forever keep increase the disk size.

**Comment by titanous** [comment]
@dottodot We're aware of the issue. Work is in progress to allow deletion of releases (#2679) and support automatic garbage collection.

**Comment by dottodot** [comment]
OK but is there anything I can do in the mean time or is it just a case of increasing the disk space.

**Comment by titanous** [comment]
Yeah, use a larger disk.

**Comment by dottodot** [comment]
OK so I've just been trying the new option to delete releases in the hope of freeing up disk space, but after deleting several release the disk usage has gone up.

This is before I started deleting
```
flynn-default                                       27.8G  8.03G    19K  /flynn-default
flynn-default/031d1f2a-01db-4e4f-a225-e847be0a771b    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/031d1f2a-01db-4e4f-a225-e847be0a771b
flynn-default/1e0761b2-6ea1-4463-ac95-97fe360e7e44   175M  8.03G   175M  /var/lib/flynn/volumes/zfs/mnt/1e0761b2-6ea1-4463-ac95-97fe360e7e44
flynn-default/241f5a98-4858-4012-9f3c-378192f611b6    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/241f5a98-4858-4012-9f3c-378192f611b6
flynn-default/4baa6b8f-bb9b-4db5-8756-829670bb489f    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/4baa6b8f-bb9b-4db5-8756-829670bb489f
flynn-default/509b6428-e855-44f7-b14a-17de1d1786ad    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/509b6428-e855-44f7-b14a-17de1d1786ad
flynn-default/5df07117-00e7-4dd1-8407-3ced2722d208    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/5df07117-00e7-4dd1-8407-3ced2722d208
flynn-default/71f771d2-dbed-4978-8ff5-b8c597aad2f7    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/71f771d2-dbed-4978-8ff5-b8c597aad2f7
flynn-default/74d66f47-a631-4837-a027-ee5f2178d545  10.4G  8.03G  10.4G  /var/lib/flynn/volumes/zfs/mnt/74d66f47-a631-4837-a027-ee5f2178d545
flynn-default/7dfde359-1a20-4bc7-90e5-a41d95a1ab1f  17.3G  8.03G  17.3G  /var/lib/flynn/volumes/zfs/mnt/7dfde359-1a20-4bc7-90e5-a41d95a1ab1f
flynn-default/94b90df4-7a3d-4ee4-81b7-75e281c20ded    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/94b90df4-7a3d-4ee4-81b7-75e281c20ded
flynn-default/99cc7f10-5750-4327-a632-d17f9ae77c0b    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/99cc7f10-5750-4327-a632-d17f9ae77c0b
flynn-default/9ee507b8-3409-42f9-9c52-eda6155d3805    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/9ee507b8-3409-42f9-9c52-eda6155d3805
flynn-default/c899548c-f644-426a-b574-96cac7f67483   406K  8.03G   406K  /var/lib/flynn/volumes/zfs/mnt/c899548c-f644-426a-b574-96cac7f67483
flynn-default/cf69eef8-1938-4958-b376-8f945ed40985    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/cf69eef8-1938-4958-b376-8f945ed40985
flynn-default/d8a1d0a8-f010-413a-8542-125f16ca7e3e    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/d8a1d0a8-f010-413a-8542-125f16ca7e3e
flynn-default/ed5182ce-efe4-4cd9-8ffb-66c481211782    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/ed5182ce-efe4-4cd9-8ffb-66c481211782
flynn-default/f7cd0c77-659b-4cb9-a282-eabf03168483    19K  8.03G    19K  /var/lib/flynn/volumes/zfs/mnt/f7cd0c77-659b-4cb9-a282-eabf03168483
```
this is after
```
NAME                                                 USED  AVAIL  REFER  MOUNTPOINT
flynn-default                                       29.8G  6.07G    19K  /flynn-default
flynn-default/031d1f2a-01db-4e4f-a225-e847be0a771b    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/031d1f2a-01db-4e4f-a225-e847be0a771b
flynn-default/1e0761b2-6ea1-4463-ac95-97fe360e7e44   175M  6.07G   175M  /var/lib/flynn/volumes/zfs/mnt/1e0761b2-6ea1-4463-ac95-97fe360e7e44
flynn-default/241f5a98-4858-4012-9f3c-378192f611b6    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/241f5a98-4858-4012-9f3c-378192f611b6
flynn-default/4baa6b8f-bb9b-4db5-8756-829670bb489f    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/4baa6b8f-bb9b-4db5-8756-829670bb489f
flynn-default/509b6428-e855-44f7-b14a-17de1d1786ad    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/509b6428-e855-44f7-b14a-17de1d1786ad
flynn-default/5df07117-00e7-4dd1-8407-3ced2722d208    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/5df07117-00e7-4dd1-8407-3ced2722d208
flynn-default/71f771d2-dbed-4978-8ff5-b8c597aad2f7    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/71f771d2-dbed-4978-8ff5-b8c597aad2f7
flynn-default/74d66f47-a631-4837-a027-ee5f2178d545  12.3G  6.07G  12.3G  /var/lib/flynn/volumes/zfs/mnt/74d66f47-a631-4837-a027-ee5f2178d545
flynn-default/7dfde359-1a20-4bc7-90e5-a41d95a1ab1f  17.3G  6.07G  17.3G  /var/lib/flynn/volumes/zfs/mnt/7dfde359-1a20-4bc7-90e5-a41d95a1ab1f
flynn-default/94b90df4-7a3d-4ee4-81b7-75e281c20ded    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/94b90df4-7a3d-4ee4-81b7-75e281c20ded
flynn-default/99cc7f10-5750-4327-a632-d17f9ae77c0b    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/99cc7f10-5750-4327-a632-d17f9ae77c0b
flynn-default/9ee507b8-3409-42f9-9c52-eda6155d3805    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/9ee507b8-3409-42f9-9c52-eda6155d3805
flynn-default/c899548c-f644-426a-b574-96cac7f67483   406K  6.07G   406K  /var/lib/flynn/volumes/zfs/mnt/c899548c-f644-426a-b574-96cac7f67483
flynn-default/cf69eef8-1938-4958-b376-8f945ed40985    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/cf69eef8-1938-4958-b376-8f945ed40985
flynn-default/d8a1d0a8-f010-413a-8542-125f16ca7e3e    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/d8a1d0a8-f010-413a-8542-125f16ca7e3e
flynn-default/ed5182ce-efe4-4cd9-8ffb-66c481211782    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/ed5182ce-efe4-4cd9-8ffb-66c481211782
flynn-default/f7cd0c77-659b-4cb9-a282-eabf03168483    19K  6.07G    19K  /var/lib/flynn/volumes/zfs/mnt/f7cd0c77-659b-4cb9-a282-eabf03168483
```

**Comment by dottodot** [comment]
So as it stands at the moment I have 3 apps running with all the old releases deleted but it's using 30GB. I'm struggling to see why so much disk space is used.

**Comment by lmars** [comment]
I just tried adding and deleting a file from the blobstore and whilst the file is removed from the PostgreSQL database, the disk usage does in fact increase:

* at the start, not much used

```
$ sudo zfs list /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f
NAME                                                 USED  AVAIL  REFER  MOUNTPOINT
flynn-default/df36592a-e416-4031-b6e4-c7a93035375f  59.5M  33.1G  59.5M  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f
```

* create and upload a 100M file

```
$ dd if=/dev/urandom of=big.txt bs=1M count=100
$ curl -X PUT --data-binary @big.txt http://100.100.24.9/big.txt
```

* usage increased by ~220M

```
$ sudo zfs list /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f
NAME                                                 USED  AVAIL  REFER  MOUNTPOINT
flynn-default/df36592a-e416-4031-b6e4-c7a93035375f   283M  32.9G   283M  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f
```

* top 5 pg files by usage

```
$ sudo du --separate-dirs /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f | sort -n | tail -5
7194    /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/12173
7377    /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/16826
8778    /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/16386
131160  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/pg_xlog
145334  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/16824
```

* delete, add and then delete the same file

```
$ curl -X DELETE http://100.100.24.9/big.txt
$ curl -X PUT --data-binary @big.txt http://100.100.24.9/big.txt
$ curl -X DELETE http://100.100.24.9/big.txt
```

* usage increased by ~270M

```
$ sudo zfs list /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f                                                                
NAME                                                 USED  AVAIL  REFER  MOUNTPOINT
flynn-default/df36592a-e416-4031-b6e4-c7a93035375f   553M  32.6G   553M  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f
```

* both pg_xlog and base/16824 directories grew

```
$ sudo du --separate-dirs /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f | sort -n | tail -5
7194    /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/12173
7369    /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/16826
8778    /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/16386
245933  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/pg_xlog
283067  /var/lib/flynn/volumes-0/zfs/mnt/df36592a-e416-4031-b6e4-c7a93035375f/db/base/16824
```

Continuing to add & delete the same file leads to ever increasing disk usage.

@josephglanville @titanous any thoughts?

**Comment by titanous** [comment]
I think a [VACUUM](https://www.postgresql.org/docs/current/static/sql-vacuum.html) is necessary to actually free up space.

**Comment by titanous** [comment]
Specifically, it looks like `VACUUM FULL` should do what we want (though it won't free up the xlog entries as they are required for replication/point-in-time recovery).

**Comment by josephglanville** [comment]
Indeed, we will need to periodically run `VACUUM FULL`. We can probably just schedule it inside the blobstore app using a goroutine to only run it against the `files` table nightly or so.

@dottodot for now to reclaim the space you can run this:
```
flynn -a blobstore pg psql -- -c 'VACUUM FULL files'
```

**Comment by dottodot** [comment]
OK so I just deleted a load of releases then ran `flynn -a blobstore pg psql -- -c 'VACUUM FULL files'` but saw no change in disk usage when running `zfs list`

**Comment by josephglanville** [comment]
@dottodot I did some further investigation. A few things were of note.

First it turns out that running `VACUUM` on the `files` table will have no effect, my bad for recommending that I forgot that when using PostgreSQL large objects that the blob data will actually be stored in the `pg_largeobject` table instead.

Second, the blobstore user won't have the permission to run `VACUUM` on this table. So you will need to connect to the database as the superuser. This is unfortunately slightly convoluted right now. You start by connecting to the `postgres` system database as the `flynn` superuser. Then switch to the blobstore database, and from there finally run `VACUUM FULL pg_largeobject`
The full procedure is as follows.

1. Get the blobstore database name
```
flynn -a blobstore env get PGDATABASE
```

2. Connect to the `postgres` database as `flynn`
```
flynn -a postgres pg psql
```

3. Switch to the blobstore database
```
\c <blobstore_database_uuid>
```

4. Run the vacuum to rewrite the blobstore table
```
VACUUM FULL pg_largeobject;
```

This can take some time if you have quite a lot of slugs as it will rewrite all of the table data. You also need to have as much free space remaining as is currently in use. For this reason I suggest you avoid this procedure unless you accidentally wrote a bunch of really big binary releases and have since removed the data but space usage isn't reclaimed.

There is also another thing of note that we need to document which is the database is configured to keep 1000 WAL segments. These are the postgres intent logs that are used to stream changes to downstream slaves. We keep so many so we can ensure when bringing up a new slave the amount of time it takes between creating the basebackup and onlining the slaves we don't run out of segments.
The segment size is configured as 16MB, so 1k segments is 16GB.

What this means is that you can basically expect the database to overtime consume 16GB of un-reclaimable space. We may at some point make the amount of WAL segments configurable to make this less of a burden on very small clusters but at this time 16GB is considered a reasonable default.

**Comment by dottodot** [comment]
@josephglanville Thanks for the detailed reply, helps bring some understanding to where all the disk space is going.

**Comment by titanous** [comment]
With support for garbage collection, release deletion, the S3 backend for the blobstore, the reduction in WAL segments kept, and the upcoming switch to libcontainer that will fix #38, I think this issue has been resolved. Please open a new issue if you have any further issues.
